File: src/main/clients/openrgb/effects/SpectrumCycle.ts
import Device from '@main/clients/openrgb/client/classes/Device';
import { RGBColor } from '@main/clients/openrgb/client/classes/RGBColor';
import Client from '@main/clients/openrgb/client/client';
import { AbstractEffect } from '@main/clients/openrgb/effects/base/AbstractEffect';

class SpectrumCycle extends AbstractEffect {
  public getName(): string {
    return 'Spectrum Cycle';
  }

  protected async applyEffect(client: Client, devices: Array<Device>): Promise<void> {
    for (let offset = 0; this.isRunning; offset = (offset + 1) % 360) {
      devices.forEach((element, i) => {
        client.updateLeds(
          i,
          Array(element.leds.length).fill(RGBColor.fromHSV(offset, 1, this.brightness))
        );
      });
      await new Promise<void>((resolve) => setTimeout(resolve, 20));
    }
    this.hasFinished = true;
  }
}

export const spectrumCycle = new SpectrumCycle();

File: lib/clients/openrgb/effects/spectrum_cycle.py
from lib.clients.openrgb.effects.base.abstract_effect import AbstractEffect
from lib.utils.singleton import singleton
from lib.utils.openrgb import OpenRGBUtils


@singleton
class SpectrumCycle(AbstractEffect):
    """Spectrum Cycle efect"""

    def __init__(self):
        super().__init__("Spectrum cycle")

    def apply_effect(self):
        offset = 0
        while self.is_running:
            for dev in self.devices:
                colors = [OpenRGBUtils.from_hsv(offset, 1, 1)] * len(dev.leds)
                self._set_colors(dev, colors)
            offset = (offset + 1) % 360
            self._sleep(0.02)  # 20 ms


spectrum_cycle = SpectrumCycle()
